### PR TITLE
TF-2871 Notification almost in default notification when use OIDC in iOS

### PR DESCRIPTION
**Pull request by dab246**
## Issue

#2871 
#2900

## Root causes

- Missing `tokenEndpoint` & `scopes` in keychain to call refresh new token
- Convert ISO8601 date in swift code incorrect
- Incorrect checking of the time expiration condition `AuthenticationSSO.isExpiredTime()` in swift code
- `RefreshToken` returning null when retrieving a new token
- Time returned is of type `miliseconds` the conversion is wrong.
- Keychain cannot access when app unlocked
- Reuse token in keychain has expired

## Resolved

- IOS: Work fine
- Android: Still work fine

https://github.com/linagora/tmail-flutter/assets/80730648/9380046c-75be-4b5c-89e3-7f73af39b835

### CODE REVIEW

**Comment by tddang-linagora** [comment]
signal-2024-05-31-144840_002
Screenshot 2024-05-31 at 14 48 09
Screenshot 2024-05-31 at 14 49 55

Might want to check more on this PR

**Comment by dab246** [comment]
> signal-2024-05-31-144840_002 Screenshot 2024-05-31 at 14 48 09 Screenshot 2024-05-31 at 14 49 55
> 
> Might want to check more on this PR

Please provide clear steps. Note, you should `LOGOUT` the old app and delete the app before installing the new app so that the keychain is up to date. If possible please videotape the steps for further investigation.
- Note: I ran the test again on my computer and it is working fine.

**Comment by hoangdat** [comment]
- It works fine in my device too (IP 11 Promax - 17.5)
- [ ] but can not work well when click to notification, app can not open well, take a long time in loading screen

https://github.com/linagora/tmail-flutter/assets/6462404/cbdd1f73-e302-4cee-8cc7-2126333222d7

**Comment by dab246** [comment]
> * [ ]  It works fine in my device too (IP 11 Promax - 17.5) in short time, but after 1 hour and more, notifications all is default.

## Root cause

- Due to incorrect checking of the time expiration condition `AuthenticationSSO.isExpiredTime()` in swift code
- Due to `refreshToken` returning null when retrieving a new token
- Due the time returned is of type `miliseconds`, the conversion is wrong.


> * [ ]  but can not work well when click to notification, app can not open well, take a long time in loading screen

## Root cause

- Only happens when the app is in terminated state
- Because waiting for the result of the `notificationInteractionChannel.invokeMethod('getInitialNotificationInfo')` method has not been successful. Leading to endless loading. The cause of this wait is that `notificationInteractionChannel.invokeMethod('getInitialNotificationInfo')` is called before `notificationInteractionChannel.setMethodCallHandler()` is called.

**Comment by hoangdat** [comment]
Blocker:
- [ ] Remote notification payload not found

**Comment by dab246** [comment]
> Blocker:
> 
> * [x]  Cannot open email correctly when click on notification in app terminated

**Comment by hoangdat** [comment]
- rebase to `refactor`

reviewers:
- [ ] Android
- [ ] iOS
- [ ] web

**Comment by dab246** [comment]
> * rebase to `refactor`
> 
> reviewers:
> 
> * [ ]  Android
> * [ ]  iOS
> * [ ]  web

Rebase to the `refactor` branch is complete

File: lib/features/home/presentation/home_controller.dart [hunk 1]
```diff
@@ -90,7 +90,11 @@ class HomeController extends ReloadableController {
 
   static void downloadCallback(String id, DownloadTaskStatus status, int progress) {}
 
-  void _cleanupCache() async {
+  Future<void> _cleanupCache() async {
+    if (PlatformInfo.isIOS) {
+      _handleUserClickNotificationOnIOS();
+    }
```
Comment (hoangdat) [inline lib/features/home/presentation/home_controller.dart:96]: no better place to add this function?

**Comment by hoangdat** [comment]
- [ ] after install release  IPA, login -> then close app -> open app again -> can not access app

File: lib/features/home/presentation/home_controller.dart (hunk 1 above)
Comment (dab246) [reply lib/features/home/presentation/home_controller.dart:96]: IMO, it is the most suitable place for us to get the content from the notification when the app starts launching

**Comment by hoangdat** [comment]
- [ ] Still testing iOS

File: ios/TwakeCore/Network/Interceptor/AuthenticationInterceptor.swift [hunk 2]
```diff
@@ -41,20 +41,23 @@ class AuthenticationInterceptor: RequestInterceptor {
             }
 
             let newRefreshToken = tokenResponse.refreshToken ?? authenticationSSO.refreshToken
+            let expireTime = tokenResponse.expiresTime != nil
+                ? tokenResponse.expiresTime!.convertMillisecondsToISO8601String()
+                : nil
```
Comment (tddang-linagora) [inline ios/TwakeCore/Network/Interceptor/AuthenticationInterceptor.swift:46]:
```suggestion
            let tokenResponse.expiresTime?.convertMillisecondsToISO8601String()
```

File: lib/features/home/presentation/home_controller.dart [hunk 3]
```diff
@@ -87,9 +90,9 @@ class HomeController extends ReloadableController {
 
   static void downloadCallback(String id, DownloadTaskStatus status, int progress) {}
 
-  void _cleanupCache() async {
+  Future<void> _cleanupCache() async {
     if (PlatformInfo.isIOS) {
-      await _handleIOSDataMessage();
+      _handleUserClickNotificationOnIOS();
```
Comment (tddang-linagora) [inline lib/features/home/presentation/home_controller.dart:95]: should we `await` it?

Comment (dab246) [reply lib/features/home/presentation/home_controller.dart]: should, updated

File: ios/TwakeCore/Network/Interceptor/AuthenticationInterceptor.swift (hunk 2 above)
Comment (dab246) [reply ios/TwakeCore/Network/Interceptor/AuthenticationInterceptor.swift]: updated

**Comment by hoangdat** [comment]
- [ ] export log failed for iOS

**Comment by tddang-linagora** [comment]
- [x] iOS - Open email when app is terminated for long time (>8h) OK

**Comment by hoangdat** [comment]
- [ ] still have this error in today build
image

File: ios/TwakeCore/Utils/CoreUtils.swift [hunk 4]
```diff
@@ -8,6 +8,7 @@ class CoreUtils {
     static let NOTIFICATION_INTERACTION_CHANNEL_NAME = "notification_interaction_channel"
     static let GET_REMOTE_NOTIFICATION_PAYLOAD_METHOD_CALL_NAME = "getRemoteNotificationPayload"
     static let USER_CLICK_NOTIFICATION_IN_FOREGROUND_METHOD_CALL_NAME = "userClickNotificationInForeground"
+    static let IOS_EMAIL_NOTIFICATION_ID = "ios_email_notification_id"
```
Comment (hoangdat) [inline ios/TwakeCore/Utils/CoreUtils.swift:11]: `CURRENT_EMAIL_ID_IN_NOTIFICATION_CLICK`

File: ios/Runner/AppDelegate.swift [hunk 5]
```diff
@@ -114,6 +123,19 @@ import flutter_local_notifications
...
+              !emailId.isEmpty else {
+            return
+        }
+        TwakeLogger.shared.log(message: "handleRemoteNotificationData:emailId = \(emailId)")
+        self.saveEmailNotificationId(emailId: emailId)
+    }
+    
+    private func saveEmailNotificationId(emailId: String) {
+        UserDefaults.standard.set(emailId, forKey: CoreUtils.IOS_EMAIL_NOTIFICATION_ID)
+    }
```
Comment (hoangdat) [inline ios/Runner/AppDelegate.swift:137]: maybe we need try to store also the time of `click_event`. it will help us, not open the wrong email :)

Comment (dab246) [reply ios/Runner/AppDelegate.swift]: I don't understand what time has to do with this. IMO, we don't need it this way.

File: ios/TwakeCore/Utils/CoreUtils.swift (hunk 4 above)
Comment (dab246) [reply ios/TwakeCore/Utils/CoreUtils.swift]: done

**Comment by dab246** [comment]
> * [x]  iOS
>   
>   * click noti when app in terminated -> long time in loading in app or can not open email
>   * click noti when app in background -> ok

Fixed

**Comment by hoangdat** [comment]
- it looks like: real time update is not work well

**Comment by dab246** [comment]
> * it looks like: real time update is not work well

Tested. Still work fine. Please check again.

File: model/lib/oidc/token_oidc.dart [hunk 6]
```diff
@@ -40,8 +39,6 @@ extension TokenOIDCExtension on TokenOIDC {
   bool get isExpired {
     if (expiredTime != null) {
       final now = DateTime.now();
-      log('TokenOIDC::isExpired(): TIME_NOW: $now');
-      log('TokenOIDC::isExpired(): EXPIRED_DATE: $expiredTime');
```
Comment (hoangdat) [inline model/lib/oidc/token_oidc.dart:44]: please keep it

File: lib/features/login/data/network/interceptors/authorization_interceptors.dart [hunk 7]
```diff
@@ -48,11 +48,11 @@ class AuthorizationInterceptors extends QueuedInterceptorsWrapper {
     _token = newToken;
     _configOIDC = newConfig;
     _authenticationType = AuthenticationType.oidc;
-    log('AuthorizationInterceptors::setTokenAndAuthorityOidc: INITIAL_TOKEN = ${newToken?.token} | EXPIRED_TIME = ${newToken?.expiredTime}');
+    log('AuthorizationInterceptors::setTokenAndAuthorityOidc: TokenId = ${newToken?.tokenIdHash}');
```
Comment (hoangdat) [inline lib/features/login/data/network/interceptors/authorization_interceptors.dart:51]: please keep expired time in log

File: lib/features/login/data/network/interceptors/authorization_interceptors.dart [hunk 8]
```diff
@@ -48,11 +48,11 @@ class AuthorizationInterceptors extends QueuedInterceptorsWrapper {
...
   }
 
   void _updateNewToken(TokenOIDC newToken) {
-    log('AuthorizationInterceptors::_updateNewToken: NEW_TOKEN = ${newToken.token} | EXPIRED_TIME = ${newToken.expiredTime}');
+    log('AuthorizationInterceptors::_updateNewToken: TokenId = ${newToken.tokenIdHash}');
```
Comment (hoangdat) [inline lib/features/login/data/network/interceptors/authorization_interceptors.dart:55]: idem

File: lib/features/base/reloadable/reloadable_controller.dart [hunk 9]
```diff
@@ -26,16 +28,18 @@ import 'package:tmail_ui_user/main/utils/message_toast_utils.dart';
...
+  final UpdateAccountCacheInteractor _updateAccountCacheInteractor = Get.find<UpdateAccountCacheInteractor>();
 
   @override
   void handleFailureViewState(Failure failure) {
     if (failure is GetCredentialFailure ||
         failure is GetStoredTokenOidcFailure ||
-        failure is GetAuthenticatedAccountFailure) {
-      log('ReloadableController::handleFailureViewState(): failure: $failure');
+        failure is GetAuthenticatedAccountFailure ||
+        failure is UpdateAccountCacheFailure) {
```
Comment (hoangdat) [inline lib/features/base/reloadable/reloadable_controller.dart:38]: we dont need `goToLogin` with it

File: model/lib/oidc/token_oidc.dart (hunk 6 above)
Comment (dab246) [reply model/lib/oidc/token_oidc.dart]: done

File: lib/features/login/data/network/interceptors/authorization_interceptors.dart (hunk 7 above)
Comment (dab246) [reply lib/features/login/data/network/interceptors/authorization_interceptors.dart]: done

File: lib/features/login/data/network/interceptors/authorization_interceptors.dart (hunk 8 above)
Comment (dab246) [reply lib/features/login/data/network/interceptors/authorization_interceptors.dart]: done

File: lib/features/base/reloadable/reloadable_controller.dart (hunk 9 above)
Comment (dab246) [reply lib/features/base/reloadable/reloadable_controller.dart]: done

File: lib/features/login/domain/usecases/update_account_cache_interactor.dart [hunk 10]
```diff
@@ -2,38 +2,23 @@ import 'package:core/presentation/state/failure.dart';
...
-  final CredentialRepository _credentialRepository;
 
-  UpdateAccountCacheInteractor(
-    this._accountRepository,
-    this._credentialRepository);
+  UpdateAccountCacheInteractor(this._accountRepository);
 
-  Stream<Either<Failure, Success>> execute(Session session) async* {
+  Stream<Either<Failure, Success>> execute({required Session session, String? baseUrl}) async* {
+    final apiUrl = session.getQualifiedApiUrl(baseUrl: baseUrl);
```
Comment (hoangdat) [inline lib/features/login/domain/usecases/update_account_cache_interactor.dart:17]: move it into `try...catch`

Comment (dab246) [reply lib/features/login/domain/usecases/update_account_cache_interactor.dart]: done

**Review by hoangdat — APPROVED** [review]

**Pull request merged**